### PR TITLE
refactor: client ts migration

### PR DESCRIPTION
**Pull request by lmssiehdev**
this includes the changes mamoun made in the teams PR

### CODE REVIEW

**Comment by leia-uwu** [comment]
will hopefully review and merge this on the following days now that teams is merged
i will deal with the conflicts and errors btw

**Comment by lmssiehdev** [comment]
>i will deal with the conflicts and errors btw

I already dealt with the conflicts 💪, about the errors anything related to building/structures/stairs will probably need some work.

File: .eslintrc.json [hunk 1]
```diff
@@ -1,52 +1,51 @@
...
+    "@typescript-eslint/consistent-type-assertions": 0,
+    "@typescript-eslint/prefer-nullish-coalescing": 0,
+    "@typescript-eslint/no-this-alias": 0,
+    "@typescript-eslint/no-floating-promises": 0,
+    "@typescript-eslint/no-dynamic-delete": 0,
+    "@typescript-eslint/no-extraneous-class": 0,
+    "@typescript-eslint/no-unsafe-argument": 0,
+    "@typescript-eslint/unbound-method": 0,
+    "n/no-deprecated-api": 0,
+    "@typescript-eslint/no-non-null-asserted-optional-chain": 0
```
Comment (lmssiehdev) [inline .eslintrc.json:49]: it's not apparent from the diff but I only disabled the no-non-null-asserted-optional-chain rule.

**Comment by leia-uwu** [comment]
i just spent the last hour fixing the conflicts and all tsc errors lmao

**Comment by lmssiehdev** [comment]
💀

**Comment by lmssiehdev** [comment]
🎉

**Pull request merged**